subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java

```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
 
     @Override
     public Object get(Object key) {
+        if (key instanceof String) {
+            return attributes.get(keyFor((String) key));
+        }
         return attributes.get(key);
     }
 
     @Override
     public Object put(String key, Object value) {
         if (key == null) {
-            throw new ManifestException("The key of a manifest attribute must not be null.");
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException("The key of a manifest attribute must not be null.");
         }
         if (value == null) {
-            throw new ManifestException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+        }
+        Name name;
+        if (key.length() > 70 || key.length() * 3 > 70 && Utf8.encodedLength(key) > 70) {
+            throw new ManifestException(String.format("The Key=%s violates the Manifest spec: key should not exceed 70 bytes, actual length is %d bytes in UTF-8", key, Utf8.encodedLength(key)));
         }
         try {
-            new java.util.jar.Attributes.Name(key);
-        } catch (IllegalArgumentException e) {
-            throw new ManifestException(String.format("The Key=%s violates the Manifest spec!", key));
+            name = KNOWN_NAMES.getUnchecked(key);
```

Comment:
Great idea, however, in most cases we prefer to prove it's a performance bottleneck before optimizing with caching. I fear this may be premature optimization.
❓ I wonder if this may affect performance. I'm mostly concern about the staticness of the structure where it may keep objects around for longer than required. However, I feel like this may only be cached per build so it may be improved or on par with what we currently have. If I'm not mistaking, this should be in the "per-build" classloader and may be freed after each build. We should confirm we are not worsening the memory pressure of a build.

Reply:
Frankly speaking, I don't have lots of Manifests to test on.
I just peeked on OpenJDK implementation, and I see there's a cache there.

Then I don't like the idea that `Attributes.Name` performs `name.intern()`, so I added the cache.

> worsening the memory pressure of a build

There's limit of 1000 entries maximum. It can be lowered, however I hope people do not use randomized key names for manifests.
They would probably use 10-40 different key names maximum.